src/tree.py

```diff
@@ -0,0 +1,113 @@
+'''Module for handling suffix trees'''
+from __future__ import annotations
+from collections import deque
+from typing import Iterable
+from dataclasses import dataclass
+
+
+@dataclass
+class Knæ:
+    '''
+    Class for storing nodes. \n
+    Here Knæ is the node, parent is the Knæ one step toward the root, children is a dictionary of Knæ one step away from the root (where the key in the dictionary represents first letter in the ben of the child) and ben a tuple refering to some interval in a string.
+    '''
+    parent: Knæ | None
+    ben: tuple[int, int] | None
+    children: dict[str, Knæ] | int
+
+
+class SuffixTree:
+    '''
+    Class for storing and manipulating suffix trees. 
+    '''
+    root: Knæ
+    x: str
+
+    def __init__(self, x: str) -> None:
+        '''
+        The class SuffixTree is initialized by being created from a string x. Knæ are created so that Knæ.ben is a tuple including first index and excluding last index. This is the naive implementation. \n
+        Creating suffix tree T for string x:
+        >>> T=SuffixTree(x)
+        '''
+        self.x = x + '$'
+        self.root = Knæ(None, (0, 0), {})
+        n = len(self.x)
+        current = self.root
+        for i, _ in enumerate(x):
+            lam = 0
+            j = 0
+            while i < n:
+                if lam == current.ben[1]-current.ben[0] and current.ben[1] != n-1:
```

Comment:
If `lam` is the length of the current edge, and that edge ends at index `n-1`, you have searched to the end of a leaf-edge. If that happens, you should have hit a mismatch (sentinel against non-sentinel) before you got here. I don't think that can happen unless you have a bug somewhere else...